sw/host/opentitanlib/src/transport/hyperdebug/dfu.rs

```diff
@@ -0,0 +1,407 @@
+const DFU_STATE_DOWNLOAD_IDLE: u8 = 0x05;
+
+fn official_firmware_image() -> &'static [u8] {
+    include_bytes!(env!("hyperdebug_firmware"))
+}
+
+pub fn official_firmware_version() -> Result<&'static str> {
+    get_hyperdebug_firmware_version(official_firmware_image())
+}
+
+/// Helper method to verify that the given binary image looks like a HyperDebug firmware image.
+fn validate_firmware_image(firmware: &[u8]) -> Result<()> {
+    get_hyperdebug_firmware_version(firmware)?;
+    Ok(())
+}
+
+const EC_COOKIE: [u8; 4] = [0x99, 0x88, 0x77, 0xce];
+const EC_FIRMWARE_NAME_LEN: usize = 32;
+
+fn get_hyperdebug_firmware_version(firmware: &[u8]) -> Result<&str> {
+    let Some(pos) = firmware[0..1024]
+        .chunks(4)
+        .position(|c| c[0..4] == EC_COOKIE) else {
+            bail!(TransportError::FirmwareProgramFailed(
+                "File is not a HyperDebug firmware image".to_string()
+            ));
+        };
+    let firmware_name_field = &firmware[(pos + 1) * 4..(pos + 1) * 4 + EC_FIRMWARE_NAME_LEN];
+    let end = firmware_name_field.iter()
+        .rev()
+        .position(|b| *b != 0x00)
+        .map(|j| EC_FIRMWARE_NAME_LEN - j).unwrap_or(0);
+    Ok(std::str::from_utf8(&firmware_name_field[0..end])?)
+}
+
+/// Helper method to perform flash programming using ST's DfuSe variant of the DFU protocol.
+/// This method is used both by the `Hyperdebug` and the `HyperdebugDfu` structs.
+pub fn update_firmware(
+    usb_device: &mut UsbBackend,
+    firmware: &Option<Vec<u8>>,
+    progress: &Option<Box<dyn Fn(u32)>>,
+) -> Result<Option<Box<dyn Annotate>>> {
+    let firmware: &[u8] = if let Some(vec) = firmware.as_ref() {
+        validate_firmware_image(vec)?;
+        &vec
+    } else {
+        official_firmware_image()
+    };
+
+    let dfu_desc = scan_usb_descriptor(usb_device)?;
+
+    // Exclusively claim DFU interface, preparing for control requests.
+    usb_device.claim_interface(dfu_desc.dfu_interface)?;
+
+    if DFU_STATE_APP_IDLE != wait_for_idle(usb_device, dfu_desc.dfu_interface)? {
+        // Device is already running DFU bootloader, proceed to firmware transfer.
+        do_update_firmware(usb_device, dfu_desc, firmware, progress)?;
+        return Ok(None);
+    }
+
+    // Device is running the HyperDebug firmware, not DFU bootloader.  Ask for switch to
+    // bootloader, and then restablish USB connection.
+    let _ = (|| -> Result<()> {
+        // Switching will cause loss of USB connection, hence the function block.
+        log::info!("Requesting switch to DFU mode...");
+        usb_device.write_control(
+            rusb::request_type(
+                rusb::Direction::Out,
+                rusb::RequestType::Class,
+                rusb::Recipient::Interface,
+            ),
+            USB_DFU_DETACH,
+            1000,
+            dfu_desc.dfu_interface as u16,
+            &[],
+        )?;
+        wait_for_idle(usb_device, dfu_desc.dfu_interface)?;
+        Ok(())
+    })();
```

Comment:
Its not clear to me why this is structured as an immediately executed function block.

Reply:
That is what I had hoped to explain with the comment "Switching will cause loss of USB connection, hence the function block." at the top.

The code here asks the USB device to "detach" its ordinary interfaces and endpoints, and instead show up on the USB bus in its DFU mode.  On HyperDebug that involves resetting the CPU, and entering the ROM bootloader.

It is expected that as the code above repeatedly queries the current "State", that after a while, it will get some USB error, after which it will wait a second, and then re-scan the USB bus.  (Other DFU-capable devices could in theory switch modes without dropping the USB connection, I think.)